lib/manager/nuget/package-tree.ts

```diff
@@ -0,0 +1,90 @@
+import path from 'path';
+import Graph from 'graph-data-structure';
+import upath from 'upath';
+import xmldoc from 'xmldoc';
+import { logger } from '../../logger';
+import { globLocalFiles, readLocalFile } from '../../util/fs';
+import { regEx } from '../../util/regex';
+
+// Get all package files at any level of ancestry that depend on `packageFileName`
+export async function getDependentPackageFiles(
+  packageFileName: string
+): Promise<string[]> {
+  const packageFiles = await getAllPackageFiles();
+  const graph: ReturnType<typeof Graph> = Graph();
+
+  for (const f of packageFiles) {
+    graph.addNode(f);
+  }
+
+  for (const f of packageFiles) {
+    const packageFileContent = (await readLocalFile(f, 'utf8')).toString();
+
+    const doc = new xmldoc.XmlDocument(packageFileContent);
+    const projectReferenceAttributes = (
+      doc
+        .childrenNamed('ItemGroup')
+        .map((ig) => ig.childrenNamed('ProjectReference')) ?? []
+    )
+      .flat()
+      .map((pf) => pf.attr['Include']);
+
+    logger.debug(projectReferenceAttributes);
+    const projectReferences = projectReferenceAttributes.map((a) =>
+      upath.normalize(a)
+    );
+    const normalizedRelativeProjectReferences = projectReferences.map((r) =>
+      normalizeRelativePath(f, r)
+    );
+
+    for (const ref of normalizedRelativeProjectReferences) {
+      graph.addEdge(ref, f);
+    }
+
+    if (graph.hasCycle()) {
+      throw new Error('Circular reference detected in NuGet package files');
+    }
+  }
+
+  return recursivelyGetDependentPackageFiles(packageFileName, graph);
+}
+
+// Traverse graph and find dependent package files at any level of ancestry
+function recursivelyGetDependentPackageFiles(
+  packageFileName: string,
+  graph: ReturnType<typeof Graph>
+): string[] {
+  const dependents: string[] = graph.adjacent(packageFileName);
+
+  if (dependents.length === 0) {
+    return [];
+  }
+
+  return dependents.concat(
+    dependents.map((d) => recursivelyGetDependentPackageFiles(d, graph)).flat()
+  );
+}
+
+// Take the path relative from a package file, and make it relative from the root of the repo
+function normalizeRelativePath(
```

Comment:
This function looks a little bit ugly. 🤔

Reply:
There might be an existing function that does the same thing.

Basically, it takes a relative path from the frame of one directory, and changes it to a relative path from the frame of another.

So a package reference like `../b/b.csproj` in `/tmp/repo/a/a.csproj` would become `b/b.csproj` (relative to `/tmp/repo`).

Let me see if there's already another function that does this, otherwise I'll see what I can do to make it clearer.

Reply:
A better name might be `reframeRealtivePath`

Reply:
I cleaned this up a bit and renamed the variables to make it more clear.